### PR TITLE
NEW: @W-15951478@: Allow resourceUrls to be provided by engine from the violation and the rule

### PR DESCRIPTION
**Pull request by stephen-carter-at-sf**

### CODE REVIEW

File: packages/code-analyzer-core/src/output-format.ts [hunk 1]
```diff
@@ -191,7 +191,7 @@ function createViolationOutput(id: number, violation: Violation, runDir: string)
         endColumn: primaryLocation.getEndColumn(),
         pathLocations: [RuleType.DataFlow, RuleType.Flow].includes(rule.getType()) ? createPathLocations(codeLocations, runDir) : undefined,
         message: violation.getMessage(),
-        resources: violation.getRule().getResourceUrls()
+        resources: violation.getResourceUrls()
```
Comment (jag-j) [inline packages/code-analyzer-core/src/output-format.ts:194]: Question - I understand you are starting off with retirejs. But other engines have resource urls at the rule level right? Would it make sense to make this a either or operation right from the start? (i.e. pick urls from rule or violations, wherever they are available)

File: packages/code-analyzer-core/src/results.ts [hunk 2]
```diff
@@ -143,6 +150,10 @@ export class UnexpectedEngineErrorViolation implements Violation {
     getPrimaryLocationIndex(): number {
         return 0;
     }
+
+    getResourceUrls(): string[] {
+        return [];
```
Comment (jfeingold35) [inline packages/code-analyzer-core/src/results.ts:156]: Maybe it makes sense for this to be one our help pages?

File: packages/code-analyzer-core/src/results.ts [hunk 3]
```diff
@@ -115,6 +116,12 @@ export class ViolationImpl implements Violation {
     getPrimaryLocationIndex(): number {
         return this.apiViolation.primaryLocationIndex;
     }
+
+    getResourceUrls(): string[] {
+        const urls: string[] = this.rule.getResourceUrls();
+        return !this.apiViolation.resourceUrls ? urls :
```
Comment (jfeingold35) [inline packages/code-analyzer-core/src/results.ts:123]: I had to turn this ternary over in my head a few times before I understood what it was doing. Could you add a 1-line comment explaining that this is combining the urls at the violation- and rule-level while skipping duplicates?

File: packages/code-analyzer-core/src/results.ts (hunk 2 above)
Comment (stephen-carter-at-sf) [reply packages/code-analyzer-core/src/results.ts:156]: It depends on the engine and the situation. But core is agnostic to engines. So i'm not sure if we'll be using this resource urls right now... but maybe in the future we might.

File: packages/code-analyzer-core/src/output-format.ts (hunk 1 above)
Comment (stephen-carter-at-sf) [reply packages/code-analyzer-core/src/output-format.ts:194]: On the core side, a violation object contains a rule. So the getResourceUrls method on our ViolationImpl actually is indeed putting the rule urls first. See line 120 of packages/code-analyzer-core/src/results.ts.

File: packages/code-analyzer-core/src/results.ts (hunk 3 above)
Comment (stephen-carter-at-sf) [reply packages/code-analyzer-core/src/results.ts:123]: Sure. Done.

**Review by jag-j — APPROVED** [review]

**Pull request merged**